internal/runbits/changesummary/changesummary.go

```diff
@@ -22,49 +26,132 @@ func New(out output.Outputer) *ChangeSummary {
+	oldRequirements := make(map[string]string)
+	for _, source := range filter {
+		oldRequirements[fmt.Sprintf("%s/%s", source.Namespace, source.Name)] = *source.Version
+	}
+
+	// List additional dependencies.
+	cs.out.Notice("") // blank line
 
-	// the added (direct dependencies) of this artifact that are actually new in this project
-	addedDependencies := funk.Join(ar.Dependencies, changed.Added, funk.InnerJoin).([]artifact.ArtifactID)
-
-	cs.out.Notice("")
-	cs.out.Notice(locale.Tl(
-		"changesummary_title",
-		"[NOTICE]{{.V0}}[/RESET] includes {{.V1}} dependencies, for a combined total of {{.V2}} new dependencies.",
-		ar.Name, strconv.Itoa(len(addedDependencies)), strconv.Itoa(len(changed.Added)),
-	))
-	for i, dep := range addedDependencies {
-		depMapping, ok := artifacts[dep]
-		if !ok {
-			multilog.Error("Could not find dependency %s in artifactsMap", dep)
+	localeKey := "additional_dependencies"
+	if hasAdditionalIndirectDependencies {
+		localeKey = "additional_total_dependencies"
+	}
+	version := ""
+	if added.Version != nil {
+		version = *added.Version
+	}
+	cs.out.Notice(locale.Tr(localeKey,
+		added.Name, version, strconv.Itoa(len(directDependencies)), strconv.Itoa(len(uniqueDependencies))))
+
+	// A direct dependency list item is of the form:
+	//   ├─ name@version (X dependencies)
+	// or
+	//   └─ name@oldVersion → name@newVersion (Updated)
+	// depending on whether or not it has subdependencies, and whether or not showUpdatedPackages is
+	// `true`.
+	for i, artifactId := range directDependencies {
+		prefix := "├─"
+		if i == len(directDependencies)-1 {
+			prefix = "└─"
+		}
+		dep := artifacts[artifactId]
+
+		version := ""
+		if dep.Version != nil {
+			version = *dep.Version
+		}
+
+		subdependencies := ""
+		if numSubs := len(dependencies[dep.ArtifactID]); numSubs > 0 && hasAdditionalIndirectDependencies {
+			subdependencies = fmt.Sprintf(" ([ACTIONABLE]%s[/RESET] dependencies)", strconv.Itoa(numSubs)) // intentional leading space
+		}
+
+		item := fmt.Sprintf("[ACTIONABLE]%s@%s[/RESET]%s", dep.Name, version, subdependencies) // intentional omission of space before last %s
+		if oldVersion, exists := oldRequirements[fmt.Sprintf("%s/%s", dep.Namespace, dep.Name)]; exists && version != "" && oldVersion != version {
+			item = fmt.Sprintf("[ACTIONABLE]%s@%s[/RESET] → %s (%s)", dep.Name, oldVersion, item, locale.Tl("updated", "updated"))
+		}
+
+		cs.out.Notice(fmt.Sprintf("  [DISABLED]%s[/RESET] %s", prefix, item))
+	}
+
+	cs.out.Notice("") // blank line
+}
+
+// isDependency iterates over all artifacts and their dependencies in the given changeset, and
+// returns whether or not the given artifact is a dependency of any of those artifacts or
+// dependencies.
+func isDependency(a artifact.ArtifactID, changeset artifact.ArtifactChangeset, artifacts artifact.Map) bool {
+	for _, artifactId := range changeset.Added {
+		if artifactId == a {
```

Comment:
I think this would always match line 132? On line 37 you iterate through this same slice, and call this function with the artifact ID in that slice.

I might be missing something though, still having my morning coffee.

Reply:
When `state install <pkg>` brings in a few additional packages, each of those packages shows up in `changeset.Added`. I need to iterate through each of those packages, and then determine if a given package is named as a dependency of one of the added (or updated) packages.

It's a complex dance, but this is necessary.

Reply:
But my point is that on line 37 you iterate through artifact ID's on the same slice you are iterating on artifact ID's here. So naturally this conditional on line 132 would match, and it would just "continue" and never actually get to the rest of the code in this function?

Reply:
No, it would still check against every other dependency in `changeset.Added` and then later `changeset.Updated`.

I hope we're not talking past each other here.

Assume you've added dependencies [foo, bar, baz], and updated dependency [quux]. foo depends on bar, and bar depends on baz and quux.

Line 37 iterates over [foo, bar, baz], and for each one calls `isDependency()`. For foo, the question is: is foo a dependency of any packages in the changeset? Well the only way to find out is to iterate over every package in the changeset: [foo, bar, baz] first, then [quux]. Naturally 132 will match the first time, so we don't bother checking to see if foo is a dependency of foo. The next two iterations will show that neither bar nor baz has foo as a dependency. Then we move on to quux and determine that foo is not a dependency of quux either. Thus, foo has been independently added and is a candidate for `state install`.

Next comes bar during the line 37 iteration. `isDependency()` is called. On the first iteration we see that bar is a dependency of foo, so go ahead and return. If this wasn't the case, the second iteration would match line 132 and skip checking if bar is a dependency of bar and continue on to baz and then quux.

etc.

Sorry if I'm stating the obvious here, but I cannot see the problem you're raising.

Reply:
Thanks for sticking with me. You are indeed stating the obvious, sometimes my brain gets stuck in a loop itself, quite the irony. 🙏